File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <bits/stdc++.h>
using namespace std;

pair<int,int> south = {1,0};
pair<int,int> north = {-1,0};
pair<int,int> east = {0,1};
pair<int,int> west = {0,-1};
vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer;

    int n = park.size();
    int m = park[0].size();
    pair<int,int> pos;

    for(int i =0; i<n; i++){
        for(int j =0; j<m; j++){
            if(park[i][j] == 'S') pos = {i,j};
        }
    }


    for(auto r: routes){
        char dir = r[0];
        int cnt = r[2] - '0';

        pair<int,int> direction;
        if(dir == 'S') direction = south;
        else if(dir == 'N') direction = north;
        else if(dir == 'E') direction = east;
        else direction = west;
        pair<int,int> current_pos = pos;

        for(int i = 0; i<cnt; i++){
            pair<int,int> next_pos = {pos.first + direction.first, pos.second + direction.second};
            int a = next_pos.first;
            int b = next_pos.second;
            if(next_pos.first < 0 || next_pos.second < 0 || next_pos.first >= n || next_pos.second >= m) {
                pos = current_pos;
                break;
            }

            if(park[next_pos.first][next_pos.second] == 'X') {
                pos = current_pos;
                break;
            }

            pos = next_pos;
        }

    }

    answer.push_back(pos.first);
    answer.push_back(pos.second);



    return answer;
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.java

import java.util.HashMap;
import java.util.Map;

class Point{
  int x, y;

  public Point(int x, int y) {
    this.x = x;
    this.y = y;
  }
}
class Solution {

  Map<String, int[]> dirMap = new HashMap<>();
  public int[] solution(String[] park, String[] routes) {

    dirMap.put("E", new int[]{0,1});
    dirMap.put("W", new int[]{0,-1});
    dirMap.put("N", new int[]{-1,0});
    dirMap.put("S", new int[]{1,0});

    Point start = null;

    for(int i = 0; i<park.length; i++){
      for(int j = 0; j<park[i].length(); j++){
        if(park[i].charAt(j) == 'S'){
          start = new Point(i,j);
        }
      }
    }

    Point currentPos = new Point(start.x, start.y);
    for(String op : routes){
      String[] curOp = op.split(" ");
      
      String dir = curOp[0];
      int cnt = Integer.parseInt(curOp[1]);

      int x = currentPos.x;
      int y = currentPos.y;
      
      boolean meetsOb = false;

      for(int i = 0; i<cnt; i++){
        int nx = x + dirMap.get(dir)[0];
        int ny = y + dirMap.get(dir)[1];

        if(nx < 0 || ny < 0 || nx >= park.length || ny >= park[0].length()) {
          meetsOb = true;
          continue;
        };
        if(park[nx].charAt(ny) == 'X') {
          meetsOb = true;
          continue;
        };
        
        x = nx;
        y = ny;
      }
      if(!meetsOb) {
        currentPos.x = x;
        currentPos.y = y;
      }
    }

    return new int[]{currentPos.x, currentPos.y};
  }

}